Initial setup of consultant-js

AlexKolpa opened the pull request:
This sets up the basic version of consultant-js including:
- [x] service registration/deregistration
- [x] configuration retrieval
- [x] configuration callbacks
- [x] example service
- [x] documentation

@Magnetme/developers ready for review!

AlexKolpa committed:
Initial setup of consultant-js

TiddoLangerak commented on .babelrc:
```diff
@@ -0,0 +1,6 @@
+{
+  "presets": [
+    "es2015",
+    "stage-3"
```
IIRC these now actually skip things in `es2016`. I would strongly recommend to use [`babel-preset-env`](https://github.com/babel/babel-preset-env) instead.

TiddoLangerak commented on config.js:
```diff
@@ -0,0 +1,67 @@
+import {parseKey, keyApplies} from './key-parser';
+import atob from 'atob';
+
+const loadConfig = async (consulHost, prefix, service) => {
```
$0.02  it might be nice to used named parameters here.

TiddoLangerak commented on key-parser.js:
```diff
@@ -0,0 +1,49 @@
+		&& isSetAndEqual(key.serviceInstance, service.serviceInstance);
+}
+
+function isSetAndEqual(keyField, serviceField) {
```
It returns `true` when `keyField` is *not* set, so it should be named `isUnsetOrEqual`

TiddoLangerak commented on config.js:
```diff
@@ -0,0 +1,67 @@
+		.map(kvPair => {
+			return {[kvPair.key.key] : atob(kvPair.value)};
+		})
+		.reduce((properties, kvPair) => Object.assign(properties, kvPair), {});
```
This section could use some naming improvements/comments. I think I kinda get what it's doing, but it's not exactly easy to understand with all different usages of `key`.

TiddoLangerak commented on config.js:
```diff
@@ -0,0 +1,67 @@
+	updates.reduce((props, key) => {
+		props[key] = newConfig[key];
+		return props;
+	}, properties);
```
`Object.assign(properties, newConfig);`

TiddoLangerak commented on config.js:
```diff
@@ -0,0 +1,67 @@
+
+	const properties = await loadConfig(host, prefix, service);
+
+	const timerId = setInterval(async () => await updateConfig(properties, await loadConfig(host, prefix, service), callbacks), interval);
```
Minor, but you don't need to `await` on `updateConfig`, it doesn't return a promise.

TiddoLangerak commented on index.js:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Globally shared instance for accessing config and service. Will not be initialized until a call to `init` is made.
+ */
+export default instance;
```
:-1: for a shared global object, *especially* from library code. That introduces all sort of global state & testing mayhem. The library should just expose factory methods, it's up to users of the library to determine how they're going to pass it around in their code.

TiddoLangerak commented on package.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "consultant-js",
```
`@magnet.me/consultant-js`

TiddoLangerak commented on package.json:
```diff
@@ -0,0 +1,21 @@
+    "babel-preset-stage-3": "^6.24.1",
+    "express": "^4.15.3"
+  }
+}
```
You should also add links to the repository etc. If you run `npm init` again I think it'll automatically add them, since it sees that you're in a git repository.

TiddoLangerak commented on service.js:
```diff
@@ -0,0 +1,88 @@
+		url : `${host}/v1/agent/service/deregister/${instance.serviceInstance}`,
+		method : 'DELETE',
+		headers : {
+			['user-agent'] : 'consultant-js'
```
Very minor, but it might be worth it to extract this to a shared config.

rogierslag replied:
include the `consultant-js` version as well

TiddoLangerak commented on service.js:
```diff
@@ -0,0 +1,88 @@
+		await promisify(cb => request(req, cb));
+	}
+	catch (e) {
+		throw new ConsultantError('Could not deregister from Consul');
```
Maybe also include the actual error?

TiddoLangerak commented on service.js:
```diff
@@ -0,0 +1,88 @@
+
+	let agent = null;
+	try {
+		const res = await promisify(cb => request({uri : `${host}/v1/agent/self`, timeout : 5000}, cb));
```
Missing UA.

TiddoLangerak commented on service.js:
```diff
@@ -0,0 +1,88 @@
+		agent = JSON.parse(res.body);
+	}
+	catch (e) {
+		throw new ConsultantError('Could not retrieve agent from consul');
```
You could also include the actual error here.

TiddoLangerak commented on config.js:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	return {
+		properties,
```
I'm not a fan of a shared properties object. There are no safe guards that other code won't alter this, and it's very easy to do so accidentally in JS. Additionally, it is quite prone to subtle bugs in async code. E.g. consider this seemingly correct snippet:

```js
async function() {
    await fetch(properties.fetchUrl);
    console.log(`Succesfully fetched ${properties.fetchUrl}`);
}
```

If the `fetchUrl` has changed between the call to await and the log then the log statement won't match what actually happened. 

I would strongly suggest to instead expose a getter that creates a new fresh clone on every call. This prevents all these global state issues, and additionally it helps a lot with change detection (it allows users to see *what* has changed, not just *if* something has changed, and it also allows users to do "lazy" change detection). 


Making this a getter function would also eliminate most of the logic in `updateConfig`, since it then only needs to trigger listeners instead of updating global state.

TiddoLangerak commented on key-parser.js:
```diff
@@ -0,0 +1,50 @@
+		&& isSetAndEqual(key.instance, service.instance);
+}
+
+function isSetAndEqual(keyField, serviceField) {
```
`isUnsetOrEqual`

TiddoLangerak commented on service.js:
```diff
@@ -0,0 +1,86 @@
+	};
+
+	const body = {
+		['Id'] : identifier.instance,
```
The keys are doubly unnecessary verbose. This can just be `const body = { Id : identifier.instance, ...}`. Same goes at more places in this PR. You only need to quote when there are special characters in the string (such as in `'user-agent'`), but even then you don't need the `[]`. Those are only needed for computed property names.

rogierslag commented on config.js:
```diff
@@ -0,0 +1,79 @@
+		.map(kvPair => {
+			return {[kvPair.key.key] : atob(kvPair.value)};
+		})
+		.reduce((properties, kvPair) => Object.assign(properties, kvPair), {});
```
a small comment with the general response format might be nice

rogierslag commented on config.js:
```diff
@@ -0,0 +1,79 @@
+	updates.reduce((props, key) => {
+		props[key] = newConfig[key];
+		return props;
+	}, properties);
```
what about deleted properties? Are these now set to undefined instead of `null`?

rogierslag replied:
I should read before commenting ;)

rogierslag commented on config.js:
```diff
@@ -0,0 +1,79 @@
+	}
+};
+
+export default async function ({consulHost, service, prefix, interval = 500}) {
```
protip, name the exported function since its name shows up in debugging sourcemaps, which makes your life easier

rogierslag replied:
Plus twice per second is a lot for config updates if you ask me

AlexKolpa replied:
It's what we use in our Java Consultant as well.

TiddoLangerak reviewed (changes requested):
Nice work! I do really think the global state should be eliminated before this can be merged. Other than that also some minor code-organizational issues:
- I would put all the code in a `src` folder. The root folder is probably also going to get a `README`, `.eslintrc`, etc. etc., it's nice to keep the source separated
- It might be nice to add a precompiled library bundle as well. Allthough this can also be done in a follow-up of course.

TiddoLangerak commented on config.js:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	return {
+		properties,
```
I'm not a fan of a shared properties object. There are no safe guards that other code won't alter this, and it's very easy to do so accidentally in JS. Additionally, it is quite prone to subtle bugs in async code. E.g. consider this seemingly correct snippet:

```js
async function() {
    await fetch(properties.fetchUrl);
    console.log(`Succesfully fetched ${properties.fetchUrl}`);
}
```

If the `fetchUrl` has changed between the call to await and the log then the log statement won't match what actually happened. 

I would strongly suggest to instead expose a getter that creates a new fresh clone on every call. This prevents all these global state issues, and additionally it helps a lot with change detection (it allows users to see *what* has changed, not just *if* something has changed, and it also allows users to do "lazy" change detection). 


Making this a getter function would also eliminate most of the logic in `updateConfig`, since it then only needs to trigger listeners instead of updating global state.

AlexKolpa replied:
Good point, I'll wrap it in a get function

TiddoLangerak commented:
Github & Rebasing messed up my comments a bit, so I think you should also still take a look at those that are marked outdated.

rogierslag commented on index.js:
```diff
@@ -0,0 +1,81 @@
+/**
+ * Globally shared instance for accessing config and service. Will not be initialized until a call to `init` is made.
+ */
+export default instance;
```
isnt it easier to return this from the init and mutate it from there? Now there is a global lock on it

rogierslag commented on key-parser.js:
```diff
@@ -0,0 +1,50 @@
+
+		tail = tail.substring(specifierEnd + 1);
+
+		return Object.assign({serviceName, key : tail}, ...specifiers);
```
this case would really benefit from some example response which is being transformed.

rogierslag commented on key-parser.js:
```diff
@@ -0,0 +1,50 @@
+	return !keyField || keyField === serviceField;
+}
+
+export {parseKey, keyApplies};
```
newline

rogierslag commented on key-parser.js:
```diff
@@ -0,0 +1,50 @@
+
+function isSetAndEqual(keyField, serviceField) {
+	return !keyField || keyField === serviceField;
+}
```
this return true is keyField is not set, which is quite different from the method name

rogierslag commented on service.js:
```diff
@@ -0,0 +1,86 @@
+const deregister = async (host, instance) => {
+	const req = {
+		url : `${host}/v1/agent/service/deregister/${instance.instance}`,
+		method : 'DELETE',
```
you can also just issue a `GET`. Consul is not that critical

AlexKolpa replied:
Just because you can, doesn't mean you should. Additionally I remember something about nginx liking to retry GETs, which might not be what you want here.

rogierslag commented on service.js:
```diff
@@ -0,0 +1,86 @@
+		dataCenter : dataCenter || process.env.SERVICE_DC || agent.Config.Datacenter,
+		host : host || process.env.SERVICE_HOST || agent.Config.NodeName,
+		instance : instance || process.env.SERVICE_INSTANCE || uuid.v4()
+	};
```
you might want to extract this to an `serviceIdentifier.js` utility. It happens in more places and really looks alike

rogierslag commented on service.js:
```diff
@@ -0,0 +1,86 @@
+	}
+};
+
+export default async function register({service:{port, name, dataCenter, host, instance}, healthCheckPath, healthCheckInterval = 10, consulHost}) {
```
newline a bit here for readability

rogierslag commented on service.js:
```diff
@@ -0,0 +1,86 @@
+		['Id'] : identifier.instance,
+		['Name'] : identifier.name,
+		['Address'] : identifier.host,
+		['Port'] : port,
```
what happens if this is not a number? Does consul reject silently or fail hard? Maybe check the number and throw before we get here?

rogierslag commented on service.js:
```diff
@@ -0,0 +1,86 @@
+		['Port'] : port,
+	};
+
+	if (healthCheckPath && healthCheckInterval > 0) {
```
I happen to know the interval is in seconds, but its not documented here (or anywhere)

rogierslag commented on service.js:
```diff
@@ -0,0 +1,86 @@
+		Object.assign(body, {
+			['Check'] : {
+				['HTTP'] : `http://${identifier.host}:${port}${healthCheckPath}`,
+				['Interval'] : healthCheckInterval
```
Do you need the array object syntax or not? If yes, please comment (otherwise some will clear it up later). If no, plz refactor

rogierslag reviewed (approved):
NICE!

AlexKolpa committed:
Remove global state from consultant

It's been moved to the example to show how users of the lib could use it.

AlexKolpa committed:
Fix request specification

Use proper key specification. Also move some properties to a shared properties module.

AlexKolpa committed:
Remove shared properties object

Instead, it is now copied to a new object for each caller and callback, to avoid accidentally mutating its state.

AlexKolpa committed:
Move identifier construction to a separate module

This makes it easier to re-use between service and config

AlexKolpa committed:
Improve readability key parsing

Extracted the specifier construction to a separate function to allow for a bit more documentation. Also fixed function naming.

AlexKolpa committed:
Extract config body parsing to function

AlexKolpa committed:
Use babel-preset-env

Also switch over to non-deprecated uuid module.

AlexKolpa committed:
Update package.json

Include GitHub links

AlexKolpa commented:
@TiddoLangerak @rogierslag thanks for all the feedback! I believe I've covered all of it now!

rogierslag commented on example/example-server.js:
```diff
@@ -0,0 +1,85 @@
+const init = async (configuration) => {
+	const {identifier, deregister} = await service(configuration);
+
+	Object.assign(instance, {identifier, deregister});
```
why not do `const instance = Object.assign({}, {identifier, deregister});`. That was the instance it local to the method

rogierslag commented on example/example-server.js:
```diff
@@ -0,0 +1,85 @@
+	return instance;
+};
+
+init({service : {name, port}, healthCheckPath : '/_health', prefix : 'config'}).then(() => {
```
then you get the instance here as a result

rogierslag commented on example/example-server.js:
```diff
@@ -0,0 +1,85 @@
+let server = null;
+
+app.get('/_health', (req, res) => {
+	res.sendStatus(200);
```
_technically_ you are not sending a body, which you should for a 200...

AlexKolpa replied:
>res.sendStatus(200); // equivalent to res.status(200).send('OK')

rogierslag replied:
TIL

rogierslag commented on example/example-server.js:
```diff
@@ -0,0 +1,85 @@
+	await stop();
+
+	process.exit();
+});
```
And this is a event handler, so you can spawn this multiple times. If there are no running things anymore, the process should exit aynway

_alternative: just keep a list of all instances, and stop all of those_

rogierslag commented on package.json:
```diff
@@ -0,0 +1,31 @@
+  "name": "@magnet.me/consultant-js",
+  "version": "1.0.0",
+  "description": "A library to register services to Consul and listen for property changes.",
+  "main": "index.js",
```
`src/index.js`

TiddoLangerak replied:
This is now `src/index.js` :smile:

rogierslag commented on package.json:
```diff
@@ -0,0 +1,31 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Magnetme/consultant-js.git"
```
~~easier to only list https. People without certificates installed cannot access the ssh version~~ nvm thats what you did

rogierslag commented on src/agent.js:
```diff
@@ -0,0 +1,20 @@
+	catch (e) {
+		throw new ConsultantError(`Could not retrieve agent from consul: ${e}`);
+	}
+}
```
newline

rogierslag commented on src/config.js:
```diff
@@ -0,0 +1,91 @@
+
+const loadConfig = async (consulHost, prefix, service) => {
+	const prefixUri = prefix ? `${prefix}/` : '';
+	const response = await promisify(cb => request({
```
since you are on promisifying http requests, why not make it just execute the method directly, so the syntax becomes

```javascript
const response = await promiseRequest({...params});
```

AlexKolpa replied:
I could. But then in a few weeks/months I have another older lib I want to use with callbacks, and I'd need to revert everything again. I'd argue this is quite readable and abstracting it away further will only make it overly specific.

rogierslag commented on src/config.js:
```diff
@@ -0,0 +1,91 @@
+		.filter(kvPair => kvPair.key.key && keyApplies(kvPair.key, service))
+		.map(kvPair => {
+			return {[kvPair.key.key] : atob(kvPair.value)};
+		})
```
you can inline this (add parentheses)

AlexKolpa replied:
Nope. You're returning an object, so you need to do a full expansion. Otherwise it sees the brackets as part of the lambda.

rogierslag replied:
You can do

`.map(kvPair => ({[kvPaid.key.key]: atob(kvPair.value)}))`

AlexKolpa replied:
I have to say I find the readability of that a bit lower than the expanded version. If you want I can add it, but I'd rather have readability over brevity.

rogierslag commented on src/config.js:
```diff
@@ -0,0 +1,91 @@
+
+	const liveProperties = await loadConfig(consulHost, prefix, identifier);
+
+	let timerId = 0;
```
why assign this dummy value?

rogierslag replied:
You could inadvertently clear the wrong timer as a result

rogierslag commented on src/service.js:
```diff
@@ -0,0 +1,73 @@
+	consulHost = consulHost || process.env.CONSUL_HOST || properties.defaultHost;
+
+	if (!Number(service.port)) {
+		throw new ConsultantError('service.port was not defined properly');
```
ok now I'm getting annoying, but also check whether the port matches `0 < port < 65536`

AlexKolpa replied:
Hahaha, valid point though!

AlexKolpa committed:
Clean up example

Pass the instance to the server startup.

AlexKolpa committed:
Add readme

Also fix package name.

wouter-willems commented on README.md:
```diff
@@ -0,0 +1,42 @@
+### Specifying an alternative Consul address
+Consultant defaults Consul's REST API address to `http://localhost:8500`. If you wish to specify an alternative address to Consul's REST API, you can do so by setting the `consulHost` field.
+
+Or alternatively you can also define the this through an environment variable:
```
typo: 'the this'
